### PR TITLE
Document that `RawWakerVTable` functions must be thread-safe.

### PR DESCRIPTION
**Pull request by kpreid**
Also add some intra-doc links and more high-level explanation of how `Waker` is used, while I'm here.

Context: https://internals.rust-lang.org/t/thread-safety-of-rawwakervtables/17126

### CODE REVIEW

File: library/core/src/task/wake.rs [hunk 1]
```diff
@@ -172,15 +184,15 @@ pub struct Context<'a> {
 }
 
 impl<'a> Context<'a> {
-    /// Create a new `Context` from a `&Waker`.
+    /// Create a new [`Context`] from a [`&Waker`](Waker).
```
Comment (steffahn) [inline library/core/src/task/wake.rs:187]: As far as I’m aware, I believe there’s commonly *no* links to the type itself in an inherent method. I.e. the `Context` being a link isn’t really useful at all.
```suggestion
    /// Create a new `Context` from a [`&Waker`](Waker).
```

I’m also adding comments to other places where a link links to the place we’re already at.

File: library/core/src/task/wake.rs [hunk 2]
```diff
@@ -202,7 +214,18 @@ impl fmt::Debug for Context<'_> {
 /// This handle encapsulates a [`RawWaker`] instance, which defines the
 /// executor-specific wakeup behavior.
 ///
-/// Implements [`Clone`], [`Send`], and [`Sync`].
+/// The typical life of a [`Waker`] is that it is constructed by an executor, wrapped in a
```
Comment (steffahn) [inline library/core/src/task/wake.rs:217]:
```suggestion
/// The typical life of a `Waker` is that it is constructed by an executor, wrapped in a
```

File: library/core/src/task/wake.rs [hunk 3]
```diff
@@ -219,18 +242,21 @@ unsafe impl Sync for Waker {}
 impl Waker {
     /// Wake up the task associated with this `Waker`.
     ///
-    /// As long as the runtime keeps running and the task is not finished, it is
-    /// guaranteed that each invocation of `wake` (or `wake_by_ref`) will be followed
-    /// by at least one `poll` of the task to which this `Waker` belongs. This makes
+    /// As long as the executor keeps running and the task is not finished, it is
+    /// guaranteed that each invocation of [`wake()`](Self::wake) (or
+    /// [`wake_by_ref()`](Self::wake_by_ref)) will be followed by at least one
+    /// [`poll()`] of the task to which this [`Waker`] belongs. This makes
```
Comment (steffahn) [inline library/core/src/task/wake.rs:248]:
```suggestion
    /// guaranteed that each invocation of `wake()` (or
    /// [`wake_by_ref()`](Self::wake_by_ref)) will be followed by at least one
    /// [`poll()`] of the task to which this `Waker` belongs. This makes
```

File: library/core/src/task/wake.rs [hunk 4]
```diff
@@ -263,7 +289,7 @@ impl Waker {
         unsafe { (self.waker.vtable.wake_by_ref)(self.waker.data) }
     }
 
-    /// Returns `true` if this `Waker` and another `Waker` have awoken the same task.
+    /// Returns `true` if this `Waker` and another [`Waker`] would awake the same task.
```
Comment (steffahn) [inline library/core/src/task/wake.rs:292]:
```suggestion
    /// Returns `true` if this `Waker` and another `Waker` would awake the same task.
```

File: library/core/src/task/wake.rs (hunk 1 above)
Comment (kpreid) [reply library/core/src/task/wake.rs]: All gone. Sorry — I actually noticed they were being excluded but not soon enough to not have made some of those changes.

Personally, I'd rather live in a world where all references to a symbol are machine-readable (whether or not they actually _become links_) but I recognize there isn't much value in trying to achieve that given current tooling (e.g. rust-analyzer does not rename in docs).

File: library/core/src/task/wake.rs [hunk 5]
```diff
@@ -71,6 +71,12 @@ impl RawWaker {
 /// pointer of a properly constructed [`RawWaker`] object from inside the
 /// [`RawWaker`] implementation. Calling one of the contained functions using
 /// any other `data` pointer will cause undefined behavior.
+///
+/// These functions must all be thread-safe (even though [`RawWaker`] is
+/// <code>\![Send] + \![Sync]</code>)
```
Comment (Nilstrieb) [inline library/core/src/task/wake.rs:76]:
```suggestion
/// `\![Send] + \![Sync]`)
```
I think the escapes can be deleted as well, though I'm not sure.

Comment (steffahn) [reply library/core/src/task/wake.rs:76]: links don't work in ``` ``` ``s.

Comment (Nilstrieb) [reply library/core/src/task/wake.rs:76]: Ah, then it makes sense.

Comment (kpreid) [reply library/core/src/task/wake.rs:76]: And the `\!` escapes too are needed to prevent interpretation as an image embed.

**Comment by scottmcm** [comment]
I don't know enough about tasks/futures to be a good reviewer here.

r? libs

**Comment by thomcc** [comment]
Hmm, can someone on @rust-lang/wg-async double-check this?

**Review by thomcc — APPROVED** [review]
This looks fine to me. I kind of would like to have someone from wg-async sign off on it, just to make sure there isn't anything new this is stating subtly

**Comment by thomcc** [comment]
After a reread through our APIs and docs I'm convinced this is fine and doesn't make any new guarantees.

@bors r+ rollup

**Comment by tmandry** [comment]
Thanks for the ping and for the PR, I agree it looks good!

**Pull request merged**